File: splitwise/go/splitwise_service.go
package main

import (
	"fmt"
	"math"
	"sync"
)

// SplitwiseService is a singleton managing the system concurrently
type SplitwiseService struct {
	mu       sync.RWMutex // Protects concurrent access to maps
	Users    map[string]*User
	Groups   map[string]*Group
	// Balances maps [UserA_ID][UserB_ID] = Amount that User A owes to User B.
	Balances map[string]map[string]float64
}

var (
	instance *SplitwiseService
	once     sync.Once
)

// GetSplitwiseService securely initializes and returns the singleton instance
func GetSplitwiseService() *SplitwiseService {
	once.Do(func() {
		instance = &SplitwiseService{
			Users:    make(map[string]*User),
			Groups:   make(map[string]*Group),
			Balances: make(map[string]map[string]float64),
		}
	})
	return instance
}

// AddUser securely registers a new user in the system
func (s *SplitwiseService) AddUser(u *User) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.Users[u.ID] = u
	s.Balances[u.ID] = make(map[string]float64)
}

// AddGroup securely registers a new group
func (s *SplitwiseService) AddGroup(g *Group) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.Groups[g.ID] = g
}

// AddExpense concurrently processes and splits a group expense
func (s *SplitwiseService) AddExpense(groupID string, expense *Expense) error {
	s.mu.Lock()
	group, exists := s.Groups[groupID]
	if !exists {
		s.mu.Unlock()
		return fmt.Errorf("group %s does not exist", groupID)
	}
	s.mu.Unlock()

	// Route the specific calculation based on split types natively before locking
	switch expense.Splits[0].(type) {
	case *EqualSplit:
		// Divide the total amount equally amongst all participants
		amountPerUser := expense.Amount / float64(len(expense.Splits))
		amountPerUser = math.Round(amountPerUser*100) / 100
		for _, split := range expense.Splits {
			split.SetAmount(amountPerUser)
		}
	case *PercentSplit:
		// Calculate precise monetary amounts based on percentage shares
		for _, rawSplit := range expense.Splits {
			percentSplit := rawSplit.(*PercentSplit)
			amount := (expense.Amount * percentSplit.Percent) / 100.0
			percentSplit.SetAmount(math.Round(amount*100) / 100)
		}
	case *ExactSplit:
		total := 0.0
		for _, split := range expense.Splits {
			total += split.GetAmount()
		}
		if total != expense.Amount {
			return fmt.Errorf("exact splits total does not match expense amount")
		}
	}

	// Update Balances globally locking the system graph
	s.mu.Lock()
	defer s.mu.Unlock()

	group.Expenses = append(group.Expenses, expense)
	paidBy := expense.PaidBy.ID

	for _, split := range expense.Splits {
		splitUser := split.GetUser().ID
		if paidBy == splitUser {
			continue // Prevent users from owing themselves
		}
		// Adjust bi-directional debt edges
		s.Balances[splitUser][paidBy] += split.GetAmount()
		s.Balances[paidBy][splitUser] -= split.GetAmount()
	}

	return nil
}

// SettleBalance securely zeros out mutual debits
func (s *SplitwiseService) SettleBalance(userA, userB string) (*Transaction, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	amountOwed := s.Balances[userA][userB]
	if amountOwed == 0 {
		return nil, fmt.Errorf("no balance to settle between %s and %s", userA, userB)
	}

	var sender, receiver string
	var amount float64

	if amountOwed > 0 {
		sender, receiver, amount = userA, userB, amountOwed
	} else {
		sender, receiver, amount = userB, userA, -amountOwed
	}

	// Reset balances iteratively
	s.Balances[userA][userB] = 0
	s.Balances[userB][userA] = 0

	return &Transaction{
		ID:       fmt.Sprintf("TRX-%s-%s", sender, receiver),
		Sender:   s.Users[sender],
		Receiver: s.Users[receiver],
		Amount:   amount,
	}, nil
}

// PrintBalances securely traces all nodes
func (s *SplitwiseService) PrintBalances() {
	s.mu.RLock()
	defer s.mu.RUnlock()

	fmt.Println("--- System Balances ---")
	for userA, owesMap := range s.Balances {
		for userB, amount := range owesMap {
			if amount > 0 {
				fmt.Printf("%s owes %s: $%.2f\n", s.Users[userA].Name, s.Users[userB].Name, amount)
			}
		}
	}
	fmt.Println("-----------------------")
}

File: splitwise/java-without-multithreading/SplitwiseService.java
import java.util.HashMap;
import java.util.Map;

public class SplitwiseService {
    private static SplitwiseService instance;

    private Map<String, User> users;
    private Map<String, Group> groups;
    // Balances maps [UserA_ID][UserB_ID] = Amount that User A owes to User B.
    private Map<String, Map<String, Double>> balances;

    private SplitwiseService() {
        users = new HashMap<>();
        groups = new HashMap<>();
        balances = new HashMap<>();
    }

    public static SplitwiseService getInstance() {
        if (instance == null) {
            instance = new SplitwiseService();
        }
        return instance;
    }

    public void addUser(User u) {
        users.put(u.getId(), u);
        balances.put(u.getId(), new HashMap<>());
    }

    public void addGroup(Group g) {
        groups.put(g.getId(), g);
    }

    public void addExpense(String groupId, Expense expense) throws Exception {
        Group group = groups.get(groupId);
        if (group == null) {
            throw new Exception("group " + groupId + " does not exist");
        }

        Split typeCheck = expense.getSplits().get(0);

        if (typeCheck instanceof EqualSplit) {
            double amountPerUser = expense.getAmount() / expense.getSplits().size();
            amountPerUser = Math.round(amountPerUser * 100.0) / 100.0;
            for (Split split : expense.getSplits()) {
                split.setAmount(amountPerUser);
            }
        } else if (typeCheck instanceof PercentSplit) {
            for (Split split : expense.getSplits()) {
                PercentSplit percentSplit = (PercentSplit) split;
                double amount = (expense.getAmount() * percentSplit.getPercent()) / 100.0;
                split.setAmount(Math.round(amount * 100.0) / 100.0);
            }
        } else if (typeCheck instanceof ExactSplit) {
            double total = 0.0;
            for (Split split : expense.getSplits()) {
                total += split.getAmount();
            }
            if (Math.abs(total - expense.getAmount()) > 0.01) {
                throw new Exception("exact splits total does not match expense amount");
            }
        }

        group.addExpense(expense);

        String paidBy = expense.getPaidBy().getId();
        for (Split split : expense.getSplits()) {
            String splitUser = split.getUser().getId();
            if (paidBy.equals(splitUser)) {
                continue;
            }

            Map<String, Double> splitUserBalances = balances.get(splitUser);
            splitUserBalances.put(paidBy, splitUserBalances.getOrDefault(paidBy, 0.0) + split.getAmount());

            Map<String, Double> paidByBalances = balances.get(paidBy);
            paidByBalances.put(splitUser, paidByBalances.getOrDefault(splitUser, 0.0) - split.getAmount());
        }
    }

    public Transaction settleBalance(String userA, String userB) throws Exception {
        double amountOwed = balances.get(userA).getOrDefault(userB, 0.0);
        if (amountOwed == 0) {
            throw new Exception("no balance to settle between " + userA + " and " + userB);
        }

        String sender, receiver;
        double amount;

        if (amountOwed > 0) {
            sender = userA;
            receiver = userB;
            amount = amountOwed;
        } else {
            sender = userB;
            receiver = userA;
            amount = -amountOwed;
        }

        balances.get(userA).put(userB, 0.0);
        balances.get(userB).put(userA, 0.0);

        return new Transaction(
            "TRX-" + sender + "-" + receiver,
            users.get(sender),
            users.get(receiver),
            amount
        );
    }

    public void printBalances() {
        System.out.println("--- System Balances ---");
        for (String userA : balances.keySet()) {
            Map<String, Double> owesMap = balances.get(userA);
            for (String userB : owesMap.keySet()) {
                double amount = owesMap.get(userB);
                if (amount > 0) {
                    System.out.printf("%s owes %s: $%.2f\n", users.get(userA).getName(), users.get(userB).getName(), amount);
                }
            }
        }
        System.out.println("-----------------------");
    }
}
